### DISCUSSION TITLE
Using micromark-extension-gfm-table 1.0.6 with remark-lint reports cell padding issues that cannot be resolved

### DISCUSSION THREAD

**Discussion by Trott** [micromark/micromark#146] [Q&A] [answered]
I have tracked an issue to an upgrade from `micromark-extension-gfm-table` 1.0.5 to 1.0.6. See https://github.com/nodejs/node/pull/47995#issuecomment-1546813989 for more information. I'm hoping someone else is already experiencing this issue and has reported a bug, but I'm not seeing anything. `micromark-extension-gfm-table` is not a direct dependency so the bug might be elsewhere, but I'm guessing not.

```
└─┬ remark-preset-lint-node@3.4.0
  └─┬ remark-gfm@3.0.1
    └─┬ micromark-extension-gfm@2.0.2
      └── micromark-extension-gfm-table@1.0.6
```

I can try to reduce the issue to something that contains less tooling and other dependencies to reduce the likelihood that something else is the source of the issue. But I don't want to duplicate effort that someone else may have already done.

So I guess my questions are: Is anyone experiencing something similar? Is the source of the issue already obvious from what I've reported and linked to? Or do I need to simplify things to make it a better bug report and/or provide more information?

**Comment by ChristianMurphy** [comment]
Hey @Trott! 👋 
Sorry you ran into a spot of trouble.

> I'm hoping someone else is already experiencing this issue and has reported a bug

This is the first I'm hearing of the issue.

> I can try to reduce the issue to something that contains less tooling and other dependencies to reduce the likelihood that something else is the source of the issue. 

Assistance with isolating the issue would be appreciated!

The unified/remark/micromark team has traditionally used sandboxes to help isolate the issue, and have some starters:
https://codesandbox.io/s/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel
https://stackblitz.com/github/remarkjs/.github/tree/main/sandbox-templates/remark-with-parcel

but other approaches to isolating the issue are also welcome!

> Is the source of the issue already obvious from what I've reported and linked to? 

Not to me.
I know https://github.com/micromark/micromark-extension-gfm-table/releases/tag/1.0.6 was a significant refactor, but the test suite is fairly comprehensive and is still passing 🤔

**Reply by wooorm** [reply]
Haven't noticed any failures in other packages that use it either. 
Also we do use remark-lint in all projects, which do have tables, and nothing fails

**Reply by Trott** [reply]
OK, I've got it reduced to three small files to demonstrate the issue.

package.json

```json
{
  "type": "module",
  "scripts": {
    "test": "remark test.md --use ./index.js -fq"
  },
  "dependencies": {
    "remark-gfm": "^3.0.1",
    "remark-lint-table-cell-padding": "^4.1.2"
  },
  "devDependencies": {
    "remark-cli": "^11.0.0"
  }
}
```

index.js

```js
import remarkGfm from "remark-gfm";
import remarkLintTableCellPadding from "remark-lint-table-cell-padding";

const plugins = [
  remarkGfm,
  remarkLintTableCellPadding,
];

const remarkPresetLintNode = { plugins };

export default remarkPresetLintNode;
```

test.md

```md
| Meaning       | Info string  |
| ------------- | ------------ |
| Powershell    | `powershell` |
| Shell Session | `console`    |
```

If you do `npm install && npm test` you get this output:

```console
test.md
  3:17  warning  Cell should be padded with 1 space, not 2  table-cell-padding  remark-lint
  4:16  warning  Cell should be padded                      table-cell-padding  remark-lint

⚠ 2 warnings
```

If you downgrade to micromark-extenstion-gfm-table 1.0.5, the markdown file passes linting, which I believe is the correct result.

Should I move this to a bug report in micromark-extension-gfm-table? Or is there more to do (such as further simplification of the demonstration code, or more work to confirm the problem is in micromark-extension-gfm-table and not a dependency or dependent of it)?

**Reply by Trott** [reply]
If I change the markdown test file to this inconsistently-padded markdown table, it passes linting with 1.0.6 and (correctly) fails with 1.0.5:

```markdown
| Meaning       | Info string  |
| ------------- | ------------ |
| Powershell    |`powershell` |
| Shell Sessio  | `console`    |
```

**Reply by wooorm** [reply]
Thanks! Can reproduce it in `mdast-util-gfm-table`.
The problem is that the positional info changed by one character:

```
| Meaning       | Info string  |
^^^^^^^^^^^^^^^^^ first cell in 1.0.5 (1:1-1:18)
^^^^^^^^^^^^^^^^ first cell in 1.0.6 (1:1-1:17)
                 ^^^^^^^^^^^^^^^ second cell in 1.0.5 (1:18-1:33)
                ^^^^^^^^^^^^^^^^ second cell in 1.0.6 (1:17-1:33)
| ------------- | ------------ |
```

So the question is, who “owns” the pipe?

Two examples that might help discuss this:

```
a|b
-|-
```

a|b
-|-

So “outer” pipes are not needed, one pipe for two columns is fine.
Pipes aren’t even needed for one column:

```
a
:-
```

a
:-

---

I don’t have strong opinions on whether the first cell should get up to 2 pipes, and latter ones up to 1 (old behavior), or the earlier cells get up to 1 pipe and the last cell can have 2 (new behavior).
I am not aware of any spec ([GFM](https://github.github.com/gfm/#tables-extension-), [mdast](https://github.com/syntax-tree/mdast#tablecell)) discussing this.
So I don‘t think this change is bad (or particularly good).

**Reply by wooorm** [reply, marked as answer]
https://github.com/remarkjs/remark-lint/commit/639271aed95fa579623f385bade4939a9c70e959
